### ISSUE TITLE
Parse errors will be ignored on optional symbols when using either '?' or '*'

### ISSUE THREAD

**Issue by silkentrance** [pegjs/pegjs#410]
See also #262, however, here we do not have any labelled top level symbols, rather a top level symbol expects one or none sub level symbol to be present.

So given the following grammar

```
Unit
    = comments:Comment*
      options:(UnitLangOption / UnitNamespaceOption)+
      translations:Translation+
      NL*
      EOF
    {
        return {
            type     : 'unit',
            location : location(),
            comments : comments,
            options: options,
            translations: translations
        };
    }


Translation
    = NL*
      comments:Comment*
      id:TranslationId
      context:VariationContext?
      variations:Variation+
    {
        return {
            type     : 'translation',
            location : location(),
            id: id,
            context: context,
            comments : comments,
            variations: variations
        };
    }


Variation
    = NL*
      comments:Comment*
      context:VariationContext?
      option:VariationPluralOption?
      NL*
      text:Text+
      NL
    {
        return {
            type     : 'variation',
            location : location(),
            comments : comments,
            context: context,
            options: [option],
            text: text
        };
    }


Comment "comment"
    = NL* '#' !SYM_DIR value:$CHAR* NL
    {
        return {
            type     : 'comment',
            location : location(),
            value    : value
        };
    }


// FIXME:id
TranslationId
    = NL* '#' SYM_ID WS* value:$ID WS* NL
    {
        return {
            type: 'id',
            location: location(),
            value: value
        };
    }

VariationContext
    = NL* '#' SYM_CTX WS* value:$QNAME WS* NL
    {
        return {
            type: 'context',
            location: location(),
            value: value
        };
    }


// FIME:text
Text "text"
    = value:CHAR+ NL


// Options

UnitLangOption
    = NL* '#' SYM_OPT WS* key:SYM_LANG WS+ value:$LOCALEID WS* NL
    {
        return {
            type: 'option',
            location: location(),
            key: key,
            value: value
        };
    }


UnitNamespaceOption
    = NL* '#' SYM_OPT WS* key:SYM_NS WS+ value:$(NSNAME / NSDEFAULT) WS* NL
    {
        return {
            type: 'option',
            location: location(),
            key: key,
            value: value
        };
    }


VariationPluralOption
    = NL* '#' SYM_OPT WS* key:SYM_PLURAL value:CARDINALITY? WS* NL
    {
        return {
            type: 'option',
            location: location(),
            key: key,
            value: value
        };
    }


CARDINALITY "cardinality"
    = WS+ value:[0-9]+
    {
        return value;
    }

// Terminals

//FIXME:ID
ID "id"
    = $CHAR+


SYM_DIR
    = SYM_ID
    / SYM_AUT
    / SYM_OPT
    / SYM_CTX


SYM_ID "!"
    = '!'


SYM_AUT "~"
    = '~'


SYM_OPT "="
    = '='


SYM_NS "ns"
    = 'ns'


SYM_CTX "context"
    = '@'


SYM_PLURAL "plural"
    = 'plural'


SYM_LANG "lang"
    = 'lang'


LOCALEID "localeid"
    = [a-z][a-z] ('-' [A-Z]+ [a-zA-Z])? ('-' [A-Z][A-Z])?


NSDEFAULT
    = 'translation'


QNAME "qname"
    = [a-zA-Z0-9]+


NSNAME "nsname"
    = [a-zA-Z]+ [a-zA-Z0-9_]*


CHAR
    = [^\r\n\f\\]


WS "ws"
    = ' '
    / '\t'


NL "nl"
    = '\n'
    / '\r\n'
    / '\r'
    / '\f'


EOF "eof"
    = !.
```

and the following input

```
# comment
#= lang de

#! tid1.as
#@

variation1
```

and the following cli frontend for parsing

```
#!/usr/bin/env /usr/bin/node
var fs = require('fs');
var parser = require('./ypo');
var f = fs.readFileSync(process.argv[2]);

try
{
    var unit = parser.parse(f.toString());
    console.log(unit);
    console.log('---------> translations');
    console.log(unit.translations);
    console.log('---------> variations');
    console.log(unit.translations[0].variations);
}
catch (err)
{
    console.log(err);
}
```

where the grammar needs to be compiled to a parser ``ypo.js`` in the same directory as the cli frontend,

parsing the input file will not fail on the malformed ``#@`` context directive.

Removing the ``?`` after ``context:VariationContext`` in

```
Translation
    = NL*
      comments:Comment*
      id:TranslationId
      context:VariationContext
      variations:Variation+
    {
        return {
            type     : 'translation',
            location : location(),
            id: id,
            context: context,
            comments : comments,
            variations: variations
        };
    }
``` 

will cause the parser to report the expected error, namely that the ``qname`` rule did not match.

Or am I doing something wrong here?

**Comment by silkentrance** [comment]
Also, replacing the ``?`` by ``+`` instead of removing it will cause the expected SyntaxError to be thrown.

Perhaps it would be best to not quietly ignore any errors when using either ``*`` or ``?``, since a parse error is a parser error and it should always be reported, even if the "compound" symbol is optional or can occur multiple times.

**Comment by silkentrance** [comment]
Similarly, the LOCALEID when malformed, will also not cause the expected error, yet, with a bit of indirection and restructuring this will work as expected. Gosh, perhaps I only need to restructure the whole grammar so that the errors pop up where I would expect them to. Consider this closed for the time being :dancer:

**Comment by silkentrance** [comment]
Must reopen, refactoring the grammar did not help, e.g. having the following rules changed/added

```
VariationContext "context"
    = NL* '#' SYM_CTX WS* value:CONTEXTID
    {
        return {
            type: 'context',
            location: location(),
            value: value
        };
    }

CONTEXTID "contextid"
    = WS value:QNAME EOL 
    {   
        return value;
    }   

EOL "end of line"
    = WS* NL
```

will not report the error.

**Comment by silkentrance** [comment]
It might seem as if rules based on character classes are treated differently from rules made up of simple terminal strings such as

```
LCID
    = 'en-US'
    / 'en'
    / 'de-DE'
    / 'de'
```

While the latter rule will nicely detect that it had not been matched and thus will make the including rule, namely the TRANSLATIONID fail with a SyntaxError, the character class based rules will cause their including top level rule to silently fail.

**Comment by silkentrance** [comment]
Definitely in need of #248

**Comment by silkentrance** [comment]
Reopening as I am still unabled to fix the parser behavior as displayed in https://github.com/pegjs/pegjs/issues/410#issuecomment-176139444

**Comment by dmajda** [comment]
@silkentrance Could you please add a minimal grammar + input triggering the issue, a description of what you expected to happen, and a description of what happened instead? Thanks.

**Comment by silkentrance** [comment]
@dmajda I have created a gist, which I should have done in the first place...

https://gist.github.com/silkentrance/bd77c42a57be9ac66603


In the ypo-minimal-notfailing.pegjs grammar, the optional Translation .. context:VariationContext? is not followed by other symbols and thus the expected SyntaxError 'contextid' expected is thrown.

However, in the ypo-minimal-failing-1.pegjs grammar, the Translation awaits additional Variation symbols to follow. Here, the malformed VariationContext, which is still optional, will not be detected and that part of the input file gets swallowed up, never showing in the output.

**Comment by dmajda** [comment]
@silkentrance Sorry, but an example consisting of 605 lines of code is not what by meant by *minimal*. Let me be more precise and throw in a working definition:

A *minimal example reproducing a problem* is a grammar + input where removing any significant part of the grammar or input will make stop it form reproducing the problem. (In a grammar, whitespace and comments are generally not considered significant.)

Now, could you please provide a minimal example according to this definition?

**Comment by silkentrance** [comment]
@dmajda I have updated the gist.

FYI

- cli.js is a simple frontend for running the grammar and outputting some "debug" information

Since this is a line based parser, we do need the other production rules such as NL and WS and EOF, so I moved these rules way to the bottom and left the relevant rules at the top

- ypo-minimal-notfailing.pegjs is @ 30 relevant lines and the relevant production rules here are ``Translation `` and ``VariationContext`` and ``CONTEXT_ID``

- ypo-minimal-failing-1.pegjs is @ 48 relevant lines in order to be able to reproduce the error and the relevant production rules here are ``Translation ``, ``Variation``, ``VariationContext`` and ``CONTEXT_ID`` and ``Text``

The extra two production rules ``Variation`` and ``Text`` are necessary for this ``error`` to be perceivable, namely ``CONTEXT_ID`` production rule failing and the parser moving forward to finally applying the ``Text`` rule.

Having analyzed this further, it actually seems as if one not only needs an option/directive to ignore errors from a specific rule but to also enforce a specific rule to fail regardless of whether the top level rule, such as ``VariationContext`` was declared optional.

So it might basically boil down to just #411.

Further complicating the grammar might do the trick as well such as ``Text`` not matching any directives, e.g. ``#!`` or ``#@`` would cause the text rule to fail. I will try this as soon as I find the time and report back whether I was able to solve the problem.

**Comment by silkentrance** [comment]
I have reworked the ``Text`` rule so that it will no longer act as a catchall and now the expected SyntaxError ``Expected contextid' shows. See ``ypo-minimal-notfailing-1.pegjs``.

I am closing this now and thanks taking the time of looking into it.

Perhaps something to be included with the #248 where it reads ``For absolute PEGJS beginners...`` :grin:.
